CI: Docker Hub rate limit failures on ARM runners due to unauthenticated pull of tonistiigi/binfmt

## Problem

The `integration (ubuntu-24.04-arm, arm64, ...)` CI jobs intermittently fail at the "Install Docker Client" step with:

```
docker: toomanyrequests: You have reached your unauthenticated pull rate limit.
https://www.docker.com/increase-rate-limit
```

This happens because `install-docker.sh` runs `docker run --privileged --rm tonistiigi/binfmt --install all` without authenticating to Docker Hub. GitHub Actions runners share IP pools, so the anonymous rate limit (100 pulls/6h per IP) is easily exhausted.

Example failure: https://github.com/cortexproject/cortex/actions/runs/24256703215/job/70830753787?pr=7384

<details>
<summary>Full build log from failed step</summary>

```
Run sudo ./.github/workflows/scripts/install-docker.sh
  sudo ./.github/workflows/scripts/install-docker.sh
  shell: /usr/bin/bash -e {0}
  env:
    GOTOOLCHAIN: local
+ VER=29.2.1
++ uname -m
+ ARCH=aarch64
+ case $ARCH in
+ DOCKER_ARCH=aarch64
+ BUILDX_ARCH=arm64
+ echo 'Installing Docker 29.2.1 for architecture: aarch64 (docker: aarch64, buildx: arm64)'
+ curl -L -o /tmp/docker-29.2.1.tgz https://download.docker.com/linux/static/stable/aarch64/docker-29.2.1.tgz
Installing Docker 29.2.1 for architecture: aarch64 (docker: aarch64, buildx: arm64)
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 72.6M  100 72.6M    0     0   283M      0 --:--:-- --:--:-- --:--:--  283M
+ tar -xz -C /tmp -f /tmp/docker-29.2.1.tgz
+ mkdir -vp /root/.docker/cli-plugins/
mkdir: created directory '/root/.docker'
mkdir: created directory '/root/.docker/cli-plugins/'
+ curl --silent -L https://github.com/docker/buildx/releases/download/v0.31.1/buildx-v0.31.1.linux-arm64
+ chmod a+x /root/.docker/cli-plugins/docker-buildx
+ mv /tmp/docker/containerd /tmp/docker/containerd-shim-runc-v2 /tmp/docker/ctr /tmp/docker/docker /tmp/docker/docker-init /tmp/docker/docker-proxy /tmp/docker/dockerd /tmp/docker/runc /usr/bin
+ docker run --privileged --rm tonistiigi/binfmt --install all
Unable to find image 'tonistiigi/binfmt:latest' locally
latest: Pulling from tonistiigi/binfmt
docker: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit

Run 'docker run --help' for more information
Error: Process completed with exit code 125.
```

</details>

## Possible Solutions

1. **Skip binfmt on ARM runners** — The `tonistiigi/binfmt --install all` step installs QEMU for cross-arch emulation. If ARM integration tests only run ARM-native containers, this step is unnecessary on ARM runners.
2. **Mirror the image to GHCR** — Push `tonistiigi/binfmt` to `ghcr.io/cortexproject/binfmt` and pull from there. GHCR has no rate limits for public images within GitHub Actions.
3. **Authenticate with Docker Hub** — Add `docker login` using a stored access token before the pull.
4. **Pin and cache the image** — Pin to a specific tag/digest and cache it as an artifact to avoid repeated pulls.

## Affected Files

- `.github/workflows/scripts/install-docker.sh`
- `.github/workflows/test-build-deploy.yml`

--- a/.github/workflows/scripts/install-docker.sh
+++ b/.github/workflows/scripts/install-docker.sh
@@ -28,4 +28,7 @@ mkdir -vp ~/.docker/cli-plugins/
 curl --silent -L "https://github.com/docker/buildx/releases/download/v0.31.1/buildx-v0.31.1.linux-$BUILDX_ARCH" > ~/.docker/cli-plugins/docker-buildx
 chmod a+x ~/.docker/cli-plugins/docker-buildx
 mv /tmp/docker/* /usr/bin
+if [ -n "$DOCKER_REGISTRY_PASSWORD" ]; then
+  docker login -u "$DOCKER_REGISTRY_USER" -p "$DOCKER_REGISTRY_PASSWORD"
+fi
 docker run --privileged --rm tonistiigi/binfmt --install all

--- a/.github/workflows/test-build-deploy.yml
+++ b/.github/workflows/test-build-deploy.yml
@@ -261,7 +261,7 @@ jobs:
         run: |
           docker pull minio/minio:RELEASE.2024-05-28T17-19-04Z
           docker pull consul:1.8.4
-          docker pull gcr.io/etcd-development/etcd:v3.4.7
+          docker pull quay.io/coreos/etcd:v3.5.29
           if [ "$TEST_TAGS" = "integration_backward_compatibility" ]; then
             docker pull quay.io/cortexproject/cortex:v1.16.1
             docker pull quay.io/cortexproject/cortex:v1.17.2

--- a/.github/workflows/test-build-deploy.yml
+++ b/.github/workflows/test-build-deploy.yml
@@ -135,6 +135,9 @@ jobs:
           git config --system --add safe.directory $GITHUB_WORKSPACE
       - name: Install Docker Client
         run: ./.github/workflows/scripts/install-docker.sh
+        env:
+          DOCKER_REGISTRY_USER: ${{secrets.DOCKER_REGISTRY_USER}}
+          DOCKER_REGISTRY_PASSWORD: ${{ secrets.DOCKER_REGISTRY_PASSWORD }}
       - name: Sym Link Expected Path to Workspace
         run: |
           mkdir -p /go/src/github.com/cortexproject/cortex
@@ -237,8 +240,6 @@ jobs:
           go-version: 1.26.2
       - name: Checkout Repo
         uses: actions/checkout@8e8c483db84b4bee98b60c0593521ed34d9990e8 # v6.0.1
-      - name: Install Docker Client
-        run: sudo ./.github/workflows/scripts/install-docker.sh
       - name: Sym Link Expected Path to Workspace
         run: |
           sudo mkdir -p /go/src/github.com/cortexproject/cortex
@@ -303,8 +304,6 @@ jobs:
     steps:
       - name: Checkout Repo
         uses: actions/checkout@8e8c483db84b4bee98b60c0593521ed34d9990e8 # v6.0.1
-      - name: Install Docker Client
-        run: sudo ./.github/workflows/scripts/install-docker.sh
       - name: Download Docker Images Artifact
         uses: actions/download-artifact@37930b1c2abaa49bbe596cd826c3c89aef350131 # v7.0.0
         with:
@@ -336,6 +335,9 @@ jobs:
           git config --system --add safe.directory $GITHUB_WORKSPACE
       - name: Install Docker Client
         run: ./.github/workflows/scripts/install-docker.sh
+        env:
+          DOCKER_REGISTRY_USER: ${{secrets.DOCKER_REGISTRY_USER}}
+          DOCKER_REGISTRY_PASSWORD: ${{ secrets.DOCKER_REGISTRY_PASSWORD }}
       - name: Sym link Expected Path to Workspace
         run: |
           mkdir -p /go/src/github.com/cortexproject/cortex

--- a/integration/e2e/images/images.go
+++ b/integration/e2e/images/images.go
@@ -10,6 +10,6 @@ var (
 	Redis      = "docker.io/redis:7.0.4-alpine"
 	Minio      = "minio/minio:RELEASE.2024-05-28T17-19-04Z"
 	Consul     = "consul:1.8.4"
-	ETCD       = "gcr.io/etcd-development/etcd:v3.4.7"
+	ETCD       = "quay.io/coreos/etcd:v3.5.29"
 	Prometheus = "quay.io/prometheus/prometheus:v3.8.1"
 )

